intermediate_source/rpc_tutorial.rst

```diff
@@ -0,0 +1,529 @@
+   might be implementing a parameter server training framework, where model
+   parameters and trainers live on different machines.
+
+
+The `torch.distributed.rpc <https://pytorch.org/docs/master/rpc.html>`__ package
+can help with the above scenarios. In case 1, `RPC <https://pytorch.org/docs/master/rpc.html#rpc>`__
+and `RRef <https://pytorch.org/docs/master/rpc.html#rref>`__ can help send data
+from one worker to another and also easily referencing remote data objects. In
+case 2, `distributed autograd <https://pytorch.org/docs/master/rpc.html#distributed-autograd-framework>`__
+and `distributed optimizer <https://pytorch.org/docs/master/rpc.html#module-torch.distributed.optim>`__
+allows executing backward and optimizer step as if it is local training. In the
+next two sections, we will demonstrate APIs of
+`torch.distributed.rpc <https://pytorch.org/docs/master/rpc.html>`__ using a
+reinforcement learning example and a language model example. Please note, this
+tutorial is not aiming at building the most accurate or efficient models to
+solve given problems, instead the main goal is to show how to use the
+`torch.distributed.rpc <https://pytorch.org/docs/master/rpc.html>`__ package to
+build distributed training applications.
+
+
+
+Distributed Reinforcement Learning using RPC and RRef
+-----------------------------------------------------
+
+This section describes steps to build a toy distributed reinforcement learning
+model using RPC to solve CartPole-v1 from `OpenAI Gym <https://gym.openai.com>`__.
+The policy code is mostly borrowed from the existing single-thread
+`example <https://github.com/pytorch/examples/blob/master/reinforcement_learning>`__
+as shown below. We will skip details of the ``Policy`` design, and focus on RPC
+usages.
+
+.. code:: python
+
+    class Policy(nn.Module):
+
+        def __init__(self):
+            super(Policy, self).__init__()
+            self.affine1 = nn.Linear(4, 128)
+            self.dropout = nn.Dropout(p=0.6)
+            self.affine2 = nn.Linear(128, 2)
+
+            self.saved_log_probs = []
+            self.rewards = []
+
+        def forward(self, x):
+            x = self.affine1(x)
+            x = self.dropout(x)
+            x = F.relu(x)
+            action_scores = self.affine2(x)
+            return F.softmax(action_scores, dim=1)
+
+Let's first prepare a helper function to call a function on a local ``RRef``. It
+might look unnecessary at the first glance, as you could simply do
+``rref.local_value().some_func(args)`` to run the target function. The reason
+for adding this helper function is because there is no way to get a reference
+of a remote value, and ``local_value`` is only available on the owner of the
+``RRef``.
+
+.. code:: python
+
+    def _call_method(method, rref, *args, **kwargs):
+        return method(rref.local_value(), *args, **kwargs)
+
+
+    def _remote_method(method, rref, *args, async_call=False, **kwargs):
```

Comment:
Should we just do `rpc_sync` to keep it simple for tutorial purposes? If we want the possibility to do async in this tutorial maybe we should clarify that a future which can be awaited on will be returned?

Reply:
the agent does use async call on multiple observers. Let me only use async here and explicitly call async and explain futures later when it is used.